onnxruntime/core/providers/cuda/tensor/scatter_nd.cc

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cuda/tensor/scatter_nd.h"
+#include "core/providers/cuda/tensor/scatter_nd_impl.h"
+#include "core/providers/cuda/shared_inc/cuda_utils.h"
+#include "core/providers/cpu/tensor/utils.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+ONNX_OPERATOR_VERSIONED_KERNEL_EX(ScatterND,
+                                  kOnnxDomain,
+                                  11, 12,
+                                  kCudaExecutionProvider,
+                                  KernelDefBuilder()
+                                      .TypeConstraint("T", DataTypeImpl::AllFixedSizeTensorTypes())
+                                      .MayInplace(0, 0),
+                                  ScatterND);
+
+ONNX_OPERATOR_KERNEL_EX(ScatterND,
+                        kOnnxDomain,
+                        13,
+                        kCudaExecutionProvider,
+                        KernelDefBuilder()
+                            .TypeConstraint("T", DataTypeImpl::AllFixedSizeTensorTypes())
+                            .MayInplace(0, 0),
+                        ScatterND);
+
+Status ScatterND::ComputeInternal(OpKernelContext* context) const {
+  const auto* input_tensor = context->Input<Tensor>(0);
+  const auto* indices_tensor = context->Input<Tensor>(1);
+  const auto* updates_tensor = context->Input<Tensor>(2);
+
+  const auto& input_shape = input_tensor->Shape();
+  const auto& indices_shape = indices_tensor->Shape();
+  const auto& updates_shape = updates_tensor->Shape();
+
+  // Validate input shapes
+  ValidateShapes(input_shape, indices_shape, updates_shape);
+
+  auto* output_tensor = context->Output(0, input_shape);
+
+  const void* input_data = input_tensor->DataRaw();
+  void* output_data = output_tensor->MutableDataRaw();
+
+  size_t element_size = input_tensor->DataType()->Size();
+
+  if (input_data != output_data) {
+    // TODO: Run benchmarks to determine if a dedicated kernel doing data copy will be faster than invoking cudaMemcpy ?
+    cudaMemcpyAsync(output_data, input_data, element_size * input_shape.Size(), cudaMemcpyDeviceToDevice);
```

Comment:
You don't have to call synchronize after this? How do we know the copy is complete?

Reply:
I thought about it - per my understanding we do not need a synchronize nor do we have to make this a blocking call as long as we use the default stream and anything  downstream to this uses the default stream because all tasks queued within the same stream are completed in order.